File: [1049]最后一块石头的重量Ⅱ/1049_last-stone-weight-ii.c
int lastStoneWeightII(int* stones, int stonesSize) {
    int sum = 0;
    for (int i = 0; i < stonesSize; i++) {
        sum += stones[i];
    }
    int target = sum / 2;
    int dp[target + 1];
    memset(dp, 0, sizeof(dp));
    dp[0] = 1;
    for (int i = 0; i < stonesSize; i++) {
        for (int j = target; j >= stones[i]; j--) {
            if (dp[j - stones[i]]) {
                dp[j] = 1;
            }
        }
    }
    for (int i = target; i >= 0; i--) {
        if (dp[i]) {
            return sum - 2 * i;
        }
    }
    return 0;
}
    
    

File: [1049]最后一块石头的重量Ⅱ/1049_last-stone-weight-ii.java
class Solution {
    public int lastStoneWeightII(int[] stones) {
        int sum = 0;
        for (int weight : stones) {
            sum += weight;
        }
        int target = sum / 2;
        boolean[] dp = new boolean[target + 1];
        dp[0] =  true;
        for (int weight : stones) {
            for (int j = target; j >= weight; j--) {
                if (dp[j - weight]) {
                    dp[j] = true;
                }
            }
        }
        for (int i = target; i >= 0; i--) {
            if (dp[i]) {
                return sum - 2 * i;
            }
        }
        return 0;
    }
}